### PR TITLE
Add Dan Zhang's auto-wipeout design docs to the user-data-cleanup docs.

### PR DESCRIPTION
**Pull request by eobrain**
Preview:  [Design][1], [Rules Extraction][2]

[1]: https://github.com/eobrain/functions-samples/blob/autowipeout-doc/user-data-cleanup/doc/design.md
[2]: https://github.com/eobrain/functions-samples/blob/autowipeout-doc/user-data-cleanup/doc/auto_rules_extraction.md

### CODE REVIEW

File: user-data-cleanup/doc/auto_rules_extraction.md [hunk 1]
```diff
@@ -0,0 +1,385 @@
+# Firebase Auto Wipeout: Auto Rules Extraction 
+[Dan Zhang][1], July 2017
+
+
+## Introduction
+
+The wipeout project is part of a Firebase project to help Firebase
```
Comment (horseno) [inline user-data-cleanup/doc/auto_rules_extraction.md]: I think we kind of decided to call the summer project Auto-wipeout. I'm totally open to better names, but I think we might want it to be consistent.

File: user-data-cleanup/doc/auto_rules_extraction.md [hunk 2]
```diff
@@ -0,0 +1,385 @@
...
+developer protect privacy of their end-users. The wipeout project is a
+cloud function library which can automatically delete user data on
+end-user account deletion. In this way, our customers (Firebase
+developers) can focus on making their apps special with a lower cost
+of writing safer apps.
+
+
+In order to make using the wipeout library closer to a “one-click”
+experience, wipeout rules could be automatically generated by
+information inferred from the authorization rules.This doc explains
```
Comment (horseno) [inline user-data-cleanup/doc/auto_rules_extraction.md]: Missing space before 'This'

File: user-data-cleanup/doc/auto_rules_extraction.md [hunk 3]
```diff
@@ -0,0 +1,385 @@
...
+>   enforced automatically at all times. Every read and write request
+>   will only be completed if your rules allow it.
+
+More detailed references can be found here: [Firebase Database Security
+Rules API][4].
+
+
+## Traversal of the rule tree
+
+The authorization rule of the RTDB is parse to be a json object which
```
Comment (horseno) [inline user-data-cleanup/doc/auto_rules_extraction.md]: parse -> parsed

File: user-data-cleanup/doc/auto_rules_extraction.md [hunk 4]
```diff
@@ -0,0 +1,385 @@
...
+naturally has a tree structure. We traverse the tree in a
+Breadth-First search for “.write” rules. Whenever we find a write
+rule, we first check the status implied by the write rule (discussed
+in "Single write rule analysis" below), and combine it with the status
+of its ancestor (discussed "Hierarchical write rules" below) to decide
+the status of the current location. If it turns out that the current
+place holds personal data, the location will be recorded in wipeout
```
Comment (horseno) [inline user-data-cleanup/doc/auto_rules_extraction.md]: "the location" is actually patterns for the location. The format has been described in the main design doc.

File: user-data-cleanup/doc/auto_rules_extraction.md [hunk 5]
```diff
@@ -0,0 +1,385 @@
...
+
+First, we discuss how to deal with a single 
+
+### Definitions:
+
+#### Variable
+
+The `$location` variable in firebase [security rules][5]. A variable
+that can be used to reference the key of a `$location` that was used
+earlier in a rule structure. For example, replacing the variable $uid
```
Comment (horseno) [inline user-data-cleanup/doc/auto_rules_extraction.md]: $uid => `$uid`

File: user-data-cleanup/doc/auto_rules_extraction.md [hunk 6]
```diff
@@ -0,0 +1,385 @@
...
+#### Path Pattern 
+
+A sequence of constant and variables defining the path from root to
+the current location of the write rule. For example, `/key/$k1/$k2` is
+a path pattern. It could match multiple places in the database with
+different values for the variables.
+
+#### Auth placeholder
+
+A special variable which could represent Firebase authentication uid of any *single* user. We use #USER_ID as the auth placeholder to avoid potential conflicts with variables of the same name since # is not a valid path component in Firebase RTDB rules.
```
Comment (horseno) [inline user-data-cleanup/doc/auto_rules_extraction.md]: #USER_ID should be in code format. Also, maybe we want to use #WIPEOUT_UID to be consistent with later examples.

File: user-data-cleanup/doc/auto_rules_extraction.md [hunk 7]
```diff
@@ -0,0 +1,385 @@
...
+Now it is clear that with EXP in DNF, each conjunctive clause maps to
+one access pattern. **In conclusion, in order to find out path
+patterns with single access pattern, we only need to find out the EXP
+with single conjunctive clause in DNF.**
+
+## An example
+
+Here is an example based on the previous definition and reasoning.
+
+Assume we have path pattern /key/$k1/$k2 , the table below shows the corresponding outcomes with different write rules.
```
Comment (horseno) [inline user-data-cleanup/doc/auto_rules_extraction.md]: /key/$k1/$k2 => `/key/$k1/$k2`

File: user-data-cleanup/doc/auto_rules_extraction.md [hunk 8]
```diff
@@ -0,0 +1,385 @@
...
+| Write Rule | Access Pattern(s) | EXP in  DNF | Results |
+|------------|-------------------|-------------|---------|
+| Auth.uid ==$k1 | /key/#W_UID/$k2 | ($k2) | Single access |
+| Auth.uid ==$k2 | /key/$k1/#W_UID | ($k1) | Single access |
+| Auth.uid ==$k1 && Auth.uid ==$k1 | /key/#W_UID/#W_UID | ($k1 && $k2) | Single access |
+| Auth.uid ==$k1 || Auth.uid ==$k1 | /key/#W_UID/$k2; | /key/$k1/#W_UID | ($k1) || ($k2) | Multiple access |
+| Auth.uid != null | - | (True) | Multiple access |
+| Auth.uid == null | - | (False) | No access |
+| Auth.uid == SOME_FIX_ID | - | (False) | No access |
+
```
Comment (horseno) [inline user-data-cleanup/doc/auto_rules_extraction.md]: The OR operator sign is treated as a column divider by markdown. Consider replacing line 146 with  | Auth.uid ==$k1 &#124; &#124; Auth.uid ==$k1 | /key/#W_UID/$k2; &#124; /key/$k1/#W_UID | ($k1) &#124;&#124; ($k2) | Multiple access |

File: user-data-cleanup/doc/auto_rules_extraction.md [hunk 9]
```diff
@@ -0,0 +1,385 @@
...
+When booleanValue is True or False, the value of EXP is fixed and the
+conjunctionList doesn’t matter, and it will be set to an empty list.
+When booleanValue is undefined, we go ahead and look at the
+conjunctionLists.
+
+### Building EXP
+
+We build EXP for write rules in a bottom-up manners. 
+
+*   Any literal with the value true or false will produce EXP = True
```
Comment (horseno) [inline user-data-cleanup/doc/auto_rules_extraction.md]: EXP = True , EXP = False should be in math format.

File: user-data-cleanup/doc/auto_rules_extraction.md [hunk 10]
```diff
@@ -0,0 +1,385 @@
...
+
+> Note: Shallower security rules override rules at deeper paths. Child
+> rules can only grant additional privileges to what parent nodes have
+> already declared. They cannot revoke a read or write privilege.
+
+Child rule can grant additional access, thus the access status of the
+child node not only depends on the child rule, but also the status of
+its parent.
+
+For example, parent /keys/$k1 with rule auth.uid == $k1 has
```
Comment (horseno) [inline user-data-cleanup/doc/auto_rules_extraction.md]: math formats

File: user-data-cleanup/doc/auto_rules_extraction.md [hunk 11]
```diff
@@ -0,0 +1,385 @@
...
+the conjunction list of parent and child rule. Recall with
+SINGLE_ACCESS, there should be only one conjunction, each literal in
+the conjunction adds a restriction to auth.uid. So we look at every
+literal in the parents conjunction and see if any restriction has been
+removed in the child rule, if so the child node has MULT_ACCESS, else
+the child node will have same SINGLE_ACCESS as its parent.
+
+Following the previous example, the parent node has conjunction [$k1]
+and the child rule has conjunction [$k2]. The restriction that
+auth.uid == $k1 has been removed in the child node this means the
```
Comment (horseno) [inline user-data-cleanup/doc/auto_rules_extraction.md]: math formats.

File: user-data-cleanup/doc/auto_rules_extraction.md [hunk 12]
```diff
@@ -0,0 +1,385 @@
...
+*   A path where user data is stored with an optional condition
+    stating any additional restriction. The path field has a
+    placeholder in it, and the placeholder will be replaced by the
+    authentication id of user at deletion.
+*   Or an exception stating location in other path objects which are
+    not user data. For example, shared folder under a user’s
+    directory.
+
+## Deletion Efficiency
+
```
Comment (horseno) [inline user-data-cleanup/doc/auto_rules_extraction.md]: Move this section to the main design doc. Also the current library only supports database search of direct children not deeper descendants at execution/deletion time.

File: user-data-cleanup/doc/auto_rules_extraction.md [hunk 13]
```diff
@@ -0,0 +1,385 @@
...
+When we do AND and OR operations of Expressions, their conditions will
+be joined by the corresponding operator. When we are calculating child
+node Access based on parent node Access and child rule Access, there’s
+only one case where conditions will be involved: parent and child rule
+both have SINGLE_ACCESS and child node have SINGLE_ACCESS. In this
+case the resulting condition of child node is the OR of the two
+conditions.
+
+## Output of auto-extraction
```
Comment (horseno) [inline user-data-cleanup/doc/auto_rules_extraction.md]: This section is outdated and should be removed. The output spec is described in the main design doc.

File: user-data-cleanup/doc/design.md [hunk 14]
```diff
@@ -0,0 +1,181 @@
+# Auto Wipeout Design
+[Dan Zhang][1], July 2017
+
+
+## Introduction
+
+The auto-wipeout project aims to help Firebase developer protect
```
Comment (horseno) [inline user-data-cleanup/doc/design.md]: developer => developers

File: user-data-cleanup/doc/design.md [hunk 15]
```diff
@@ -0,0 +1,181 @@
...
+privacy of their end-users. Auto wipeout is a Cloud Functions library
+which can automatically find and delete user data on end-user account
+deletion. It idea is inspired by the Google internal system Wipeout.
```
Comment (horseno) [inline user-data-cleanup/doc/design.md]: It => Its

File: user-data-cleanup/doc/design.md [hunk 16]
```diff
@@ -0,0 +1,181 @@
...
+In this way, Firebase customers (Firebase developers) can focus on
+making their apps special with a lower cost of writing safer apps.
+
+Developers can choose to specify user data storing patterns through
+local configurations or rely on our Auto Extraction algorithms. If no
+local configuration is found, the library takes advantages of the fact
+that Firebase security rules implies access pattern in the schemaless
```
Comment (horseno) [inline user-data-cleanup/doc/design.md]: implies => imply

File: user-data-cleanup/doc/design.md [hunk 17]
```diff
@@ -0,0 +1,181 @@
...
+real time database. It fetches security rules and runs the Auto
+Extraction algorithms to get Wipeout rules which describes user
```
Comment (horseno) [inline user-data-cleanup/doc/design.md]: describes => describe

File: user-data-cleanup/doc/design.md [hunk 18]
```diff
@@ -0,0 +1,181 @@
...
+apps is to identify user data storing location. The knowledge about
+where data is stored is crucial to effectively managing and protecting
+user data in a the schemaless back-end storing system (Fiorebase
+Realtime Database, Cloud Storage).
+
+One way to get these knowledge is to ask the developers to specify
+them. This is a good way to get accurate information so we allow
+developers to write local configurations describing user data storing
+patterns. The problem is this approach adds additional burden to the
+developers and may scary them away from using the feature.
```
Comment (horseno) [inline user-data-cleanup/doc/design.md]: scary => scare

File: user-data-cleanup/doc/design.md [hunk 19]
```diff
@@ -0,0 +1,181 @@
...
+The better alternative is to provide a service which automatically
+extracts user data storing pattern and asks the developer to verify.
+So we built Auto Wipeout with Auto Extraction which generates data
+storing pattern (Wipeout Rules) from Security Rules for RTDB.
+
+![Auto Wipeout Design](autowipeout-design.png)
+
+This Diagram shows the workflow of the current Auto Wipeout library
+works. But it can be generalized to a system which supports Cloud
+Storage and other variation of Wipepout. The generalized system is
```
Comment (horseno) [inline user-data-cleanup/doc/design.md]: Wipepout => Wipeout

File: user-data-cleanup/doc/design.md [hunk 20]
```diff
@@ -0,0 +1,181 @@
...
+described as the 3-step process:
+
+1.  Parse RTDB or Storage security rules to internal representations
+    (json object: Rule tree)
+    1.  RTDB parse json object 
+    2.  Cloud Storage/verax rules: lightweight parse generated from
```
Comment (horseno) [inline user-data-cleanup/doc/design.md]: Remove the mention of verax in public version. parse => parser

File: user-data-cleanup/doc/design.md [hunk 21]
```diff
@@ -0,0 +1,181 @@
...
+web confirmation UI is shown in the figure below:
+
+![showWIpeoutCOnfig](showWipeoutConfig.png)
+
+
+## Wipeout rules spec
+
+The wipeout rules is a list of JSON object, each of them describes a
+pattern of user data storage. When a user account is deleted, the
+library go through every config to remove any match with these
```
Comment (horseno) [inline user-data-cleanup/doc/design.md]: go => goes

File: user-data-cleanup/doc/design.md [hunk 22]
```diff
@@ -0,0 +1,181 @@
...
+patterns. A single config rule can have four fields:
+
+*   `path`: Mandatory field. A String indicating a location of user
+    data. A path can include place holder variables `#WIPEOUT_UID`
+    which will be replaced by auth.uid at execution time. It can also
+    include free variables which starts with $. A simple example path
```
Comment (horseno) [inline user-data-cleanup/doc/design.md]: starts => start

File: user-data-cleanup/doc/design.md [hunk 23]
```diff
@@ -0,0 +1,181 @@
...
+3.  All wipeout metadata is stored under the wipeout directory in the
+    database. We should check for collision of path name to make sure
+    the library doesn’t overwrite app data.
+4.  Expand Auto Wipeout to Cloud Storage.
+5.  Add Pre-hooks and post-hooks which allow developers to run
+    customized code before or after deletion.
+6.  Support roll back of deletion. Soft deletion for a fixed amount of
+    time.
+7.  Developer Confirmation Web UI could explain generated Wipeout Rule
+    with example or a fake user.
```
Comment (horseno) [inline user-data-cleanup/doc/design.md]: example => an example

**Comment by eobrain** [comment]
Thanks for the review Dan.  I've made all those changes and will merge now.

**Pull request merged**